### PR TITLE
migrateAccounts was implemented

### PR DESCRIPTION
**Pull request by luffy-choi**
## Proposed changes

* `migrateAccounts` migrates external Klaytn accoutns into KAS wallet.
  * In order to this, the accounts should be updated to `PublicKeyAccountType`.
  * Therefore, it uses caver keyring container for signing the transaction with a previous one.
  * Note that you should compose the fee delegated transaction.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/ground-x/caver-js-ext-kas/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/ground-x/caver-js-ext-kas)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/kas/wallet/wallet.js [hunk 1]
```diff
@@ -202,57 +202,90 @@ class Wallet {
         return this.apiInstances.statistics
     }
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:204]:
```suggestion
     /**
     * @type {KeyApi}
     */
    get keyApi() {
        return this.apiInstances.key
    }

    /**
     * @type {RegistrationApi}
     */
    get registrationApi() {
        return this.apiInstances.registration
    }
```

File: src/kas/wallet/wallet.js [hunk 2]
```diff
@@ -202,57 +202,90 @@ class Wallet {
...
+        // Define a new keyring container for account migration.
         let keyringContainer = new this.accountsMigration.keyringContainer()
-        const files = fs.readdirSync(dir)
-        const addresses = []
 
-        for (let i = 0; i < files.length; i++) {
-            const file = files[i]
-            const passphrase = passphrases[i]
+        // Request a set of public keys generated by KAS wallet service.
+        const keys = await this.apiInstances.key.key.keyCreation(chainId, { size: accounts.length })
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:245]:
```suggestion
        const keys = await this.keyApi.keyCreation(chainId, { size: accounts.length })
```

File: src/kas/wallet/wallet.js [hunk 3]
```diff
@@ -202,57 +202,90 @@ class Wallet {
...
-            rawTransactions.push(updateTx.getRLPEncoding())
+            // Remove the address from caver keyring container for memory optimization.
+            keyringContainer.remove(el.address)
         }
-        console.log(rawTransactions)
+
+        // Call an account registration API with the request containing a set of transaction RLP for updating account.
+        // This API will invoke transactions for RLPs if they are included.
+        // Otherwise, you can update your Klaytn's account with your own Klaytn node in advance.
+        const result = await this.apiInstances.registration(chainId, { body: req })
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:283]:
```suggestion
        const result = await this.registrationApi(chainId, { body: req })
```

File: src/kas/wallet/wallet.js [hunk 4]
```diff
@@ -202,57 +202,90 @@ class Wallet {
...
     /**
      * Migrates Klaytn accounts to KAS Wallet API. <br>
-     * await caver.kas.wallet.migarateAccounts('./keystores', ['passphrase1', 'passphrase2', 'passphrase3']) <br>
-     * @param {string} dir The directory path to migrate accounts to KAS.
-     * @param {Array.<string>} passphrases The directory path to migrate accounts to KAS.
+     * await caver.kas.wallet.migarateAccounts([{address:'{{address}}', key:'{{private_keys}}'}]) <br>
+     *
+     * @param {number} chainId Chain ID.
+     * @param {MigrationAccount[]} accounts An array of account objects migrated into KAS.
      * @return {}
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:235]:
```suggestion
     * @return {StatusResponse}
```

**Comment by jimni1222** [comment]
@luffy-choi Can you make test codes for new feature?

**Comment by jimni1222** [comment]
@luffy-choi Please rebase on dev branch, and make test codes.

File: index.js [hunk 5]
```diff
@@ -53,6 +53,7 @@ class CaverExtKAS extends Caver {
             feeDelegatedAccountUpdate: this.transaction.feeDelegatedAccountUpdate,
             createWithAccountKeyPublic: this.account.createWithAccountKeyPublic,
         }
+        this.kas.wallet.keyring = this.wallet.keyring
```
Comment (jimni1222) [inline index.js:56]: Is this essential? If not, i think it is better to guide to user to use `caver.keyringContainer.keyring`.

File: src/kas/wallet/wallet.js [hunk 6]
```diff
@@ -203,6 +204,12 @@ class Wallet {
     get statisticsApi() {
         return this.apiInstances.statistics
     }
+    /**
+     * @type {KeyApi}
+     */
+    get keyApi() {
+        return this.apiInstances.key
+    }
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:212]:
```suggestion
```
Please remove duplicated code.

File: src/kas/wallet/wallet.js [hunk 7]
```diff
@@ -218,57 +225,111 @@ class Wallet {
...
-     * @param {Array.<string>} passphrases The directory path to migrate accounts to KAS.
-     * @return {}
+     * await caver.kas.wallet.migarateAccounts([{address:'{{address}}', key:'{{private_keys}}'}]) <br>
+     *
+     * @param {number} chainId Chain ID.
+     * @param {MigrationAccount[]} accounts An array of account objects migrated into KAS.
+     * @return {StatusResponse}
      */
-    async migrateAccounts(dir, passphrases) {
+    async migrateAccounts(chainId, accounts) {
```
Comment (jimni1222) [inline src/kas/wallet/wallet.js:261]: Do we need to separately receive chainId as a parameter? If we can use `this.chainId` set by the user when initializing.

File: index.js (hunk 5 above)
Comment (jimni1222) [reply index.js:56]: Oh my bad. i understand why this is needed :)

File: test/intTest/walletIntTest.js [hunk 8]
```diff
@@ -3625,4 +3625,32 @@ describe('Wallet API service', () => {
         expect(ret.count > 0).to.be.true
         expect(ret.krn).to.equal(accountToTest.krn)
     }).timeout(500000)
+
+    // it('CAVERJS-EXT-KAS-INT-115: caver.kas.wallet.migrateAccounts should return status with a single key', async () => {
```
Comment (jimni1222) [inline test/intTest/walletIntTest.js:3629]:
```suggestion
    // it('CAVERJS-EXT-KAS-INT-233: caver.kas.wallet.migrateAccounts should return status with a single key', async () => {
```

File: test/intTest/walletIntTest.js [hunk 9]
```diff
@@ -3625,4 +3625,32 @@ describe('Wallet API service', () => {
...
+    //     const key = keyring._key._privateKey
+
+    //     const ret = await caver.kas.wallet.migrateAccounts(chainId, [{ address, key, nonce: 0 }])
+
+    //     expect(ret).not.to.be.undefined
+    //     expect(ret.status).not.to.be.undefined
+    //     expect(ret.status).to.be.equal('ok')
+    // }).timeout(500000)
+
+    // it('CAVERJS-EXT-KAS-INT-116: caver.kas.wallet.migrateAccounts should return status with multiple accounts', async () => {
```
Comment (jimni1222) [inline test/intTest/walletIntTest.js:3641]:
```suggestion
    // it('CAVERJS-EXT-KAS-INT-234: caver.kas.wallet.migrateAccounts should return status with multiple accounts', async () => {
```

File: test/wallet/testWalletMigration.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+            setMigrateStubs(kasKey, kasRegistration)
+
+            const ret = await caver.kas.wallet.migrateAccounts(chainId, [{ address, key, nonce }])
+
+            expect(ret).not.to.be.undefined
+            expect(ret.status).not.to.be.undefined
+            expect(ret.status).to.be.equal('ok')
+        })
+
+        it('CAVERJS-EXT-KAS-WALLET-170: should return status for a role based key account', async () => {
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:111]:
```suggestion
        it('CAVERJS-EXT-KAS-WALLET-204: should return status for a role based key account', async () => {
```

File: test/wallet/testWalletMigration.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                    } else {
+                        callback(null, raResult, { body: raResult })
+                    }
+                }
+            )
+            // sandbox.replace(caver.kas.wallet.keyApi, 'keyCreation', (chainId, body) => kcResult)
+            // sandbox.replace(caver.kas.wallet.registrationApi, 'registerAccount', (chainId, body) => raResult)
+        }
+
+        it('CAVERJS-EXT-KAS-WALLET-169: should return status with a single key account', async () => {
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:96]:
```suggestion
        it('CAVERJS-EXT-KAS-WALLET-203: should return status with a single key account', async () => {
```

File: test/wallet/testWalletMigration.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+            setMigrateStubs(kasKey, kasRegistration)
+
+            const ret = await caver.kas.wallet.migrateAccounts(chainId, [{ address, key, nonce }])
+
+            expect(ret).not.to.be.undefined
+            expect(ret.status).not.to.be.undefined
+            expect(ret.status).to.be.equal('ok')
+        })
+
+        it('CAVERJS-EXT-KAS-WALLET-171: should return status with a multisig key account', async () => {
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:126]:
```suggestion
        it('CAVERJS-EXT-KAS-WALLET-205: should return status with a multisig key account', async () => {
```

File: test/wallet/testWalletMigration.js [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+            setMigrateStubs(kasKey, kasRegistration)
+
+            const ret = await caver.kas.wallet.migrateAccounts(chainId, [{ address, key, nonce }])
+
+            expect(ret).not.to.be.undefined
+            expect(ret.status).not.to.be.undefined
+            expect(ret.status).to.be.equal('ok')
+        })
+
+        it('CAVERJS-EXT-KAS-WALLET-172: should return invalid key error', async () => {
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:141]:
```suggestion
        it('CAVERJS-EXT-KAS-WALLET-206: should return invalid key error', async () => {
```

File: test/wallet/testWalletMigration.js [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+                requestId: 'POST-/v2/registration/account-1607566714428099800',
+            })
+
+            const ret = await caver.kas.wallet.migrateAccounts(chainId, [{ address, key, nonce }])
+
+            expect(ret).not.to.be.undefined
+            expect(ret.code).not.to.be.undefined
+            expect(ret.code).to.be.equal(1061010)
+        })
+        it('CAVERJS-EXT-KAS-WALLET-173: should return partially failed', async () => {
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:160]:
```suggestion
        it('CAVERJS-EXT-KAS-WALLET-207: should return partially failed', async () => {
```

File: src/kas/wallet/wallet.js (hunk 7 above)
Comment (luffy-choi) [reply src/kas/wallet/wallet.js]: I deleted `chainId` from the parameters.
I refactored the function to take only one parameter.

File: test/wallet/testWalletMigration.js [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+                ) => {
+                    if (path === '/v2/key') {
+                        callback(null, kcResult, { body: kcResult })
+                    } else {
+                        callback(null, raResult, { body: raResult })
+                    }
+                }
+            )
+            // sandbox.replace(caver.kas.wallet.keyApi, 'keyCreation', (chainId, body) => kcResult)
+            // sandbox.replace(caver.kas.wallet.registrationApi, 'registerAccount', (chainId, body) => raResult)
```
Comment (jimni1222) [inline test/wallet/testWalletMigration.js:93]:
```suggestion

```

**Review by jimni1222 — APPROVED** [review]
LGTM :) Thank you for contribution !!

**Pull request merged**